### PR TITLE
WAG-1270 - Unit tests for wagtaillinkchecker

### PR DESCRIPTION
**Pull request by jtdevos**
## Summary

- Adds `tests/app/test_wagtaillinkchecker_scanner.py` — 30 tests covering `Link.message` branching, `get_url()` success/error/fallback paths, and `clean_url()`
- Adds `tests/app/test_wagtaillinkchecker_models.py` — 22 tests covering `Scan.is_finished`, `ScanLinkQuerySet` methods, the `delete_tag` signal, and `check_link_sync()` state transitions
- Surfaces one behavior worth a follow-up: `ScanLink.page_is_deleted` returns the slug string (or `None`) rather than a strict `bool` — documented in test comments, not fixed here

## Test plan

- [ ] `pytest tests/app/test_wagtaillinkchecker_scanner.py tests/app/test_wagtaillinkchecker_models.py` — 52 tests, all pass

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: website/tests/app/test_wagtaillinkchecker_scanner.py [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+            "app.wagtaillinkchecker.scanner.requests.get",
+            side_effect=requests_lib.exceptions.ConnectionError,
+        ):
+            result = get_url(
+                "https://example.com", page=None, site=site, get_full_result=True
+            )
+        assert result["error"] is True
+        assert "connecting" in result["error_message"]
+
+    def test_head_falls_back_to_get_on_invalid_schema(self):
```
Comment (jtdevos) [inline website/tests/app/test_wagtaillinkchecker_scanner.py:267]: **Phantom scenario — this test exercises a code path that cannot succeed in production.**

`InvalidSchema` is raised by `requests` when the URL *scheme* is unrecognized (e.g. `tel:`, `mailto:`). For a valid HTTPS URL like `https://example.com`, `requests.head` never raises `InvalidSchema`. And for any URL that *does* trigger it, `requests.get` on the same URL would raise the same exception — so mocking GET to return `200` after HEAD raises `InvalidSchema` is an impossible combination in production.

The docstring amplifies this: "which may succeed for schemas requests can handle" implies a realistic recovery path, but there's no URL where HEAD raises `InvalidSchema` and GET returns 200.

Consider replacing with a test that asserts the realistic outcome: when `get_url` is called with a `tel:` URL (no mocking), the recursive GET also raises `InvalidSchema` and `result["invalid_schema"]` ends up `True`.

File: website/tests/app/test_wagtaillinkchecker_models.py [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+            from app.wagtaillinkchecker.tasks import check_link_sync
+
+            check_link_sync(link.pk)
+
+        link.refresh_from_db()
+        assert link.crawled is True
+        assert link.broken is False
+        assert link.invalid is False
+
+    def test_null_page_on_scan_link_does_not_raise_attribute_error(self):
```
Comment (jtdevos) [inline website/tests/app/test_wagtaillinkchecker_models.py]: **Duplicate test — this is a strict subset of `test_successful_non_page_url_marks_scan_link_as_crawled_not_broken` (line 296).**

Both tests:
- create a `ScanLink` with `page=None`
- mock `get_url` to return the identical success payload
- call `check_link_sync`
- assert `link.crawled is True`

The `AttributeError` guard (`link.page and link.page.full_url`) is already implicitly exercised by the test at line 296 passing without exception. Removing this test loses zero behavioral coverage. Consider deleting it.

File: website/tests/app/test_wagtaillinkchecker_scanner.py [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+            "app.wagtaillinkchecker.scanner.requests.get",
+            side_effect=requests_lib.exceptions.Timeout,
+        ):
+            result = get_url(
+                "https://example.com", page=None, site=site, get_full_result=True
+            )
+        assert result["error"] is True
+        assert "Timeout" in result["error_message"]
+
+    def test_get_sets_none_status_code_when_exception_raised_before_response(self):
```
Comment (jtdevos) [inline website/tests/app/test_wagtaillinkchecker_scanner.py]: **Partial duplicate of `test_get_sets_error_on_generic_request_exception` (line 304).**

Both tests use an identical mock (`side_effect=requests_lib.exceptions.Timeout` on `requests.get`, `get_full_result=True`) and exercise the same `except requests.exceptions.RequestException` branch. They only differ in which key of the result dict they assert.

Consider merging the `status_code is None` assertion into the sibling test at line 304 and deleting this one.

File: website/tests/app/test_wagtaillinkchecker_models.py [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+@pytest.mark.django_db
+class TestScanIsFinished:
+    def test_is_finished_returns_true_when_status_is_completed(self):
+        from app.wagtaillinkchecker.models import Scan
+
+        scan = _make_scan(_make_site())
+        scan.status = Scan.Status.COMPLETED
+        assert scan.is_finished is True
+
+    def test_is_finished_returns_false_when_status_is_running(self):
```
Comment (jtdevos) [inline website/tests/app/test_wagtaillinkchecker_models.py:61]: **Brittle — relies on the model's default status instead of setting it explicitly.**

This test never sets `scan.status = Scan.Status.RUNNING`. It passes today because the default happens to be `RUNNING`, but if that default ever changes the failure is a bare `AssertionError` with no message indicating what status was expected.

Suggest adding one explicit line:
```python
scan.status = Scan.Status.RUNNING  # make the intent explicit
```

File: website/tests/app/test_wagtaillinkchecker_models.py [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+        assert broken not in qs
+
+
+# ---------------------------------------------------------------------------
+# ScanLink.page_is_deleted (no DB needed — pure property logic)
+# ---------------------------------------------------------------------------
+
+
+class TestScanLinkPageIsDeleted:
+    def test_page_is_deleted_when_both_page_deleted_and_slug_are_set(self):
```
Comment (jtdevos) [inline website/tests/app/test_wagtaillinkchecker_models.py]: **Name implies boolean `True` return, but `page_is_deleted` returns a string (the slug) when truthy.**

The property is `return self.page_deleted and self.page_slug`, which evaluates to the slug string — not `True`. The tests use `assert link.page_is_deleted` (truthiness), which is correct, but the name `test_page_is_deleted_when_both_page_deleted_and_slug_are_set` gives no signal about the non-boolean return type.

Any future assertion written as `link.page_is_deleted == True` or `link.page_is_deleted is True` would silently fail, and these tests wouldn't catch it. The NOTE comment in the class docblock acknowledges this, but the name should too. Consider renaming to `test_page_is_deleted_returns_truthy_slug_when_both_are_set`.

File: website/tests/app/test_wagtaillinkchecker_models.py [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+        assert invalid_and_broken not in qs
+
+    def test_valid_includes_non_invalid_links(self):
+        scan, working, broken, uncrawled, *_ = self._setup()
+        qs = scan.links.valid()
+        assert working in qs
+        assert broken in qs
+        assert uncrawled in qs
+
+    def test_broken_links_returns_only_valid_broken_links(self):
```
Comment (jtdevos) [inline website/tests/app/test_wagtaillinkchecker_models.py:134]: **Thin — the "only" in the name is not fully verified, and a key boundary case is missing.**

`broken_links()` is `valid().filter(broken=True)` — it has **no** `crawled=True` filter. But the fixture's `broken` link has `crawled=True`, so if someone accidentally added `.filter(crawled=True)` to `broken_links()`, this test would still pass.

To close the gap, add a broken+uncrawled link to the setup and assert it **does** appear in `broken_links()`:
```python
broken_uncrawled = _make_scan_link(scan, "https://example.com/broken-uncrawled", broken=True)
assert broken_uncrawled in scan.links.broken_links()
```

File: website/tests/app/test_wagtaillinkchecker_scanner.py [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+
+# ---------------------------------------------------------------------------
+# Link class
+# ---------------------------------------------------------------------------
+
+
+class TestLinkMessage:
+    def test_error_attribute_is_returned_directly(self):
+        """When Link.error is set, Link.message returns it verbatim without inspecting status_code."""
+        from app.wagtaillinkchecker.scanner import Link
```
Comment (jtdevos) [inline website/tests/app/test_wagtaillinkchecker_scanner.py]: **Cleanup — every test method in this file repeats the same inline import.**

There are 20+ occurrences of `from app.wagtaillinkchecker.scanner import Link` / `get_url` / `clean_url` scattered across individual test methods. If the module path ever changes, all 20+ must be updated individually.

Move them to the top of the file alongside the existing imports:
```python
from app.wagtaillinkchecker.scanner import Link, get_url, clean_url
```

File: website/tests/app/test_wagtaillinkchecker_scanner.py [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+            "app.wagtaillinkchecker.scanner.requests.get",
+            return_value=_make_response(600),
+        ):
+            result = get_url(
+                "https://example.com", page=None, site=site, get_full_result=True
+            )
+        assert result["error"] is True
+        assert "600" in result["error_message"]
+
+    def test_head_falls_back_to_get_on_connection_error(self):
```
Comment (jtdevos) [inline website/tests/app/test_wagtaillinkchecker_scanner.py:205]: **Plausible phantom — `ConnectionError` is a network-layer failure that would affect GET equally.**

`ConnectionError` is raised when the TCP connection can't be established (DNS failure, host unreachable, connection refused). These conditions apply at the OS network layer — they affect both HEAD and GET for the same host at the same instant. Mocking HEAD to fail with `ConnectionError` and GET to immediately return 200 is a scenario that can't occur in production.

The realistic HEAD→GET fallback is for servers that actively **reject** HEAD requests (which returns a 4xx, not a `ConnectionError`) — `test_head_falls_back_to_get_on_4xx_response` already covers that meaningful case.

Consider either removing this test or replacing it with one that verifies the GET fallback also returns an error when HEAD raises `ConnectionError`.

File: website/tests/app/test_wagtaillinkchecker_scanner.py [hunk 9]
```diff
@@ -0,0 +1,365 @@
...
+        with patch(
+            "app.wagtaillinkchecker.scanner.requests.get",
+            side_effect=requests_lib.exceptions.ConnectionError,
+        ):
+            result = get_url(
+                "https://example.com", page=None, site=site, get_full_result=True
+            )
+        assert result["error"] is True
+        assert "connecting" in result["error_message"]
+
```
[the remark this answers is not in the document]
Comment (scookgunngrp) [reply website/tests/app/test_wagtaillinkchecker_scanner.py:266]: Change will be made

File: website/tests/app/test_wagtaillinkchecker_scanner.py (hunk 3 above)
Comment (scookgunngrp) [reply website/tests/app/test_wagtaillinkchecker_scanner.py]: Change will be made

File: website/tests/app/test_wagtaillinkchecker_models.py (hunk 5 above)
Comment (scookgunngrp) [reply website/tests/app/test_wagtaillinkchecker_models.py]: Change will be made

File: website/tests/app/test_wagtaillinkchecker_models.py (hunk 6 above)
Comment (scookgunngrp) [reply website/tests/app/test_wagtaillinkchecker_models.py:134]: Change will be made

File: website/tests/app/test_wagtaillinkchecker_scanner.py (hunk 7 above)
Comment (scookgunngrp) [reply website/tests/app/test_wagtaillinkchecker_scanner.py]: Change will be made

File: website/tests/app/test_wagtaillinkchecker_models.py (hunk 2 above)
Comment (scookgunngrp) [reply website/tests/app/test_wagtaillinkchecker_models.py]: Change will be made

File: website/tests/app/test_wagtaillinkchecker_models.py (hunk 4 above)
Comment (scookgunngrp) [reply website/tests/app/test_wagtaillinkchecker_models.py:61]: Change will be made

File: website/tests/app/test_wagtaillinkchecker_scanner.py (hunk 8 above)
Comment (scookgunngrp) [reply website/tests/app/test_wagtaillinkchecker_scanner.py:205]: Change will be made to add first paragraph as a comment to the beginning of `test_head_falls_back_to_get_on_connection_error` and to create a separate test named `test_head_falls_back_to_get_on_connection_error_that_also_errors` that handles the realistic scenario described.

File: website/tests/app/test_wagtaillinkchecker_scanner.py (hunk 1 above)
Comment (scookgunngrp) [reply website/tests/app/test_wagtaillinkchecker_scanner.py:267]: Changes made to replace `https://example.com` with `tel:123-456-7890` in the tests: `test_head_falls_back_to_get_on_invalid_schema` and `test_get_sets_invalid_schema_flag_on_invalid_schema`.

**Comment by scookgunngrp** [comment]
Merging this PR. The changes from this PR will be reviewed as part of WAG-1270's PR ([PR 376](https://github.com/ustaxcourt/website-wagtail/pull/736))
